src/ble_linux.c

```diff
@@ -0,0 +1,497 @@
+	if (dev->bus)
+		sd_bus_unref(dev->bus);
+
+	free(dev);
+}
+
+int
+fido_ble_read(void *handle, unsigned char *buf, size_t len, int ms)
+{
+	struct ble *dev = (struct ble *)handle;
+	ssize_t r;
+	if (fido_hid_unix_wait(dev->status_fd, ms, NULL) < 0)
+		return -1;
+
+	r = read(dev->status_fd, buf, len);
+	if (r < 0)
+		return -1;
+
+	return (int)r;
+}
+
+int
+fido_ble_write(void *handle, const unsigned char *buf, size_t len)
+{
+	struct ble *dev = (struct ble *)handle;
+	sd_bus_message *send_msg;
+	int r = sd_bus_message_new_method_call(dev->bus, &send_msg, "org.bluez",
+					       dev->paths.control_point,
+					       DBUS_CHAR_IFACE, "WriteValue");
+	if (r < 0)
+		goto out;
+
+	r = sd_bus_message_append_array(send_msg, 'y', buf, len);
+	if (r < 0)
+		goto out;
+
+	r = sd_bus_message_append(send_msg, "a{sv}", 0);
+	if (r < 0)
+		goto out;
+
+	r = sd_bus_call(dev->bus, send_msg, 0, NULL, NULL);
+out:
+	sd_bus_message_unref(send_msg);
+
+	return (r >= 0) ? (int)len : -1;
+}
+
+size_t
+fido_ble_get_cp_size(fido_dev_t *d)
+{
+	return ((struct ble *)d->io_handle)->controlpoint_size;
+}
+
+
+static bool
+ble_fido_is_useable_device(const char *iface, sd_bus_message * reply, const char **name)
+{
+	int ret;
+	bool connected = false;
+	bool paired = false;
+	bool resolved = false;
+	bool has_service = false;
+
+	if (strcmp(iface, DBUS_DEV_IFACE)) {
+		return false;
+	}
+	sd_bus_message_enter_container(reply, 'a', "{sv}");
+	while (sd_bus_message_enter_container(reply, 'e', "sv") > 0) {
+		const char *propname;
+		int boolval;
+		ret = sd_bus_message_read(reply, "sv", &propname, "b", &boolval);
+		if (ret >= 0) {
+			if (!strcmp(propname, "Connected") && boolval)
+				connected = true;
+			if (!strcmp(propname, "Paired") && boolval)
+				paired = true;
+			if (!strcmp(propname, "ServicesResolved") && boolval)
+				resolved = true;
+		} else {
+			sd_bus_message_rewind(reply, 0);
+			ret = sd_bus_message_read_basic(reply, 's', &propname);
+			if (ret >= 0 && !strcmp(propname, "Name") &&
+			    name != NULL && sd_bus_message_read(reply, "v", "s", name) >= 0) {}
+			if (ret >= 0 && !strcmp(propname, "UUIDs") &&
+			   (sd_bus_message_enter_container(reply, SD_BUS_TYPE_VARIANT, "as") >= 0)) {
+				if (sd_bus_message_enter_container(reply, 'a', "s") >= 0) {
+					const char *uuid;
+					while(sd_bus_message_read_basic(reply, 's', &uuid) > 0) {
+						if (!strcasecmp(uuid, FIDO_SERVICE_UUID))
+							has_service = true;
+					}
+					sd_bus_message_exit_container(reply); /* s */
+				}
+				sd_bus_message_exit_container(reply); /* as */
+			} else {
+				sd_bus_message_skip(reply,"v");
+			}
+		}
+		sd_bus_message_exit_container(reply); /* sv */
+	}
+	sd_bus_message_exit_container(reply);  /* {sv} */
+	return connected && resolved && has_service && paired;
+}
+
+static int
+init_ble_fido_dev(fido_dev_info_t *di,
+    const char *path, const char *name)
+{
+	memset(di, 0, sizeof(*di));
+	if (asprintf(&di->path, "%s%s", FIDO_BLE_PREFIX, path) &&
```

Comment:
Per my copy of `asprintf(3)`: 

> If memory  allocation wasn't possible, or some other error occurs, these functions will return -1, and the contents of strp are undefined.

i.e. if `asprintf` fails, calling `free(di->path)` is undefined behavior. We'll need something like:

```c
if (asprintf(&di->path, "%s%s", FIDO_BLE_PREFIX, path) == -1) {
        fido_log_debug("%s: asprintf", __func__);
        di->path = NULL;
        goto fail;
}

/* <snip> */

fail:
      free(di->path);
      /* more error handling */
```